### PR TITLE
Fix #2981: Social Sharing

### LINKED ISSUE #2981: Increase App Virality
Please search for the spec "Increase App virality" > Sharing Button section for details

GitHub Issue iOS: https://github.com/brave/brave-ios/issues/2981
GitHub Issue Android: https://github.com/brave/brave-browser/issues/12210

### PR DESCRIPTION
**Pull request by soner-yuksel**
This Pull Request is implementing the changes on Brave Shields to promote sharing 

The implementation currently adding all Sharing features like creating the tracker tier and changes to educational onboarding and changes to the Brave shield screen and actually sharing with Activity Controller.

For sharing Part we are getting a screenshot of Global Stats View and sharing it with additional text. For some sharing types like whatsapp, slack, instagram etc.., we are expanding the snapshotted image and drawing text inside the image with CoreGraphics,

## Summary of Changes

This pull request fixes #2981

## Submitter Checklist:

- [x] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`

## Test Plan:
Part1:
Press Brave Shields on Top Bar
Checkout the changes on blocked row
Press the share button and checkout the share type screen with row types

Part2:
Hit one of the block count tier number on Browser
Press the share button and checkout the share type screen with row types

## Reviewer Checklist:

- [x] Issues include necessary QA labels:
  - `QA/(Yes|No)`
  - `release-notes/(include|exclude)`
  - `bug` / `enhancement`
- [x] Necessary [security reviews](https://github.com/brave/security/issues/new/choose) have taken place.
- [x] Adequate unit test coverage exists to prevent regressions.
- [x] Adequate test plan exists for QA to validate (if applicable).
- [x] Issue is assigned to a milestone (should happen at merge time).

### CODE REVIEW

File: Client/Extensions/UIViewExtensions.swift [hunk 1]
```diff
@@ -45,6 +45,15 @@ extension UIView {
 
         return screenshot(size, offset: offset, quality: quality)
     }
+   
+    /**
+     * Getting a snapshot froma  view using image renderer
```
Comment (iccub) [inline Client/Extensions/UIViewExtensions.swift]: `froma` typo

File: Client/Extensions/UIViewExtensions.swift [hunk 2]
```diff
@@ -45,6 +45,15 @@ extension UIView {
 
         return screenshot(size, offset: offset, quality: quality)
     }
+   
+    /**
```
Comment (iccub) [inline Client/Extensions/UIViewExtensions.swift]: Use `///` comment for documentation

File: Client/Extensions/UIViewExtensions.swift [hunk 3]
```diff
@@ -45,6 +45,15 @@ extension UIView {
...
+     * Getting a snapshot froma  view using image renderer
+     */
+    var snapshot: UIImage {
+        return UIGraphicsImageRenderer(size: bounds.size).image { _ in
```
Comment (iccub) [inline Client/Extensions/UIViewExtensions.swift]: `return` can be skipped

File: Client/Frontend/Browser/BrowserViewController/BrowserViewController+ProductNotification.swift [hunk 4]
```diff
@@ -11,6 +11,45 @@ import Shared
...
+    enum BenchmarkTrackerCountTier: Int, Equatable, CaseIterable {
+        case specialTier = 1000
+        case newbieExclusiveTier = 5000
+        case casualExclusiveTier = 10000
+        case regularExclusiveTier = 25000
+        case expertExclusiveTier = 75000
+        case professionalTier = 100000
+        case primeTier = 250000
+        case grandTier = 500000
+        case legendaryTier = 1000000
```
Comment (iccub) [inline Client/Frontend/Browser/BrowserViewController/BrowserViewController+ProductNotification.swift]: we can split bigger numbers by `_` to be easier to read like `1_000_000`

File: Client/Frontend/Browser/BrowserViewController/BrowserViewController+ProductNotification.swift [hunk 5]
```diff
@@ -111,6 +168,24 @@ extension BrowserViewController {
...
     
+    private func notifyTrackerAdsCount(_ count: Int, theme: Theme) {
+        let shareTrackersViewController = ShareTrackersController(theme: theme, trackingType: .trackerCountShare(count: count))
+        dismiss(animated: true)
+
+        shareTrackersViewController.actionHandler = { [weak self] action in
+            guard let self = self else { return }
+
+            switch action {
+                case .shareTheNewsTapped:
```
Comment (iccub) [inline Client/Frontend/Browser/BrowserViewController/BrowserViewController+ProductNotification.swift]: for one case maybe simple `if` check would be enough here?

File: Client/Frontend/Shields/ShieldsActivityItemSourceProvider.swift [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+        return text
+    }
+    
+    func activityViewController(_ activityViewController: UIActivityViewController, itemForActivityType activityType: UIActivity.ActivityType?) -> Any? {
+        let activityValueType = ActivityTypeValue.allCases.first(where: { $0.rawValue == activityType?.rawValue })
+        
+        if activityValueType != nil {
+            return nil
+        } else {
+            return text
```
Comment (iccub) [inline Client/Frontend/Shields/ShieldsActivityItemSourceProvider.swift]: Can do in 1 line, i reverted this condition too
`return activityValueType == nil ? text : nil`

File: Client/Frontend/Shields/ShieldsActivityItemSourceProvider.swift [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+        return image
+    }
+    
+    func activityViewController(_ activityViewController: UIActivityViewController, itemForActivityType activityType: UIActivity.ActivityType?) -> Any? {
+        let activityValueType = ActivityTypeValue.allCases.first(where: { $0.rawValue == activityType?.rawValue })
+        
+        if activityValueType != nil {
+            return imageWithText
+        } else {
+            return image
```
Comment (iccub) [inline Client/Frontend/Shields/ShieldsActivityItemSourceProvider.swift]: see my previous comment about 1 line simplification

File: Client/Frontend/Shields/SimpleShieldsView.swift [hunk 8]
```diff
@@ -59,20 +82,49 @@ class SimpleShieldsView: UIView, Themeable {
                     string: Strings.Shields.blockedCountLabel,
                     attributes: [.font: UIFont.systemFont(ofSize: 13.0)]
                 )
-                let attachment = ViewTextAttachment(view: self.infoButton)
-                string.append(NSAttributedString(attachment: attachment))
+                // Share UI only exist in locale JP
+                if Locale.current.regionCode != "JP" {
```
Comment (iccub) [inline Client/Frontend/Shields/SimpleShieldsView.swift]: shouldn't this check for JP existence, not the other way around?

**Review by iccub — CHANGES_REQUESTED** [review]
Nitpick review, i will run this code now and leave another round of feedback

File: Client/Frontend/Shields/ShieldsActivityItemSourceProvider.swift (hunk 6 above)
Comment (soner-yuksel) [reply Client/Frontend/Shields/ShieldsActivityItemSourceProvider.swift]: Both of the conditions are using ternary operation now.

File: Client/Frontend/Shields/SimpleShieldsView.swift (hunk 8 above)
Comment (soner-yuksel) [reply Client/Frontend/Shields/SimpleShieldsView.swift]: No we add the question mark for help navigation as text attachment normally. In JP locale we are changing this to new button design with share.

File: Client/Frontend/Shields/ShieldsActivityItemSourceProvider.swift [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+            }
+            statsInfoView.snp.makeConstraints {
+                $0.centerX.equalToSuperview()
+                $0.centerY.equalToSuperview().offset(15)
+                $0.height.equalToSuperview().multipliedBy(0.40)
+                $0.width.equalToSuperview().multipliedBy(0.70)
+            }
+        }
+    
+        let contentView = UIView(frame: CGRect(width: statsView.frame.width, height: statsView.frame.height + 85)).then {
```
Comment (Brandon-T) [inline Client/Frontend/Shields/ShieldsActivityItemSourceProvider.swift:49]: What is this 85?

File: Client/Frontend/Shields/ShieldsActivityItemSourceProvider.swift [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+            $0.backgroundColor = UIColor(rgb: 0x4C54D2)
```
Comment (Brandon-T) [inline Client/Frontend/Shields/ShieldsActivityItemSourceProvider.swift]: Perhaps use `#ColorLiteral`?

File: Client/Frontend/Shields/ShieldsActivityItemSourceProvider.swift [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+// MARK: - OptionalTextActivityItemSource
+
+class OptionalTextActivityItemSource: NSObject, UIActivityItemSource {
+    
+    let text: String
+    
+    weak var viewController: UIViewController?
+    
+    init(text: String) {
+        self.text = text
```
Comment (Brandon-T) [inline Client/Frontend/Shields/ShieldsActivityItemSourceProvider.swift:83]: Missing call to `super.init()` when inheriting `NSObject`

File: Client/Frontend/Shields/ShieldsActivityItemSourceProvider.swift [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+}
+
+// MARK: - ImageActivityItemSource
+
+class ImageActivityItemSource: NSObject, UIActivityItemSource {
+    let image: UIImage
+    let imageWithText: UIImage
+    
+    init(image: UIImage, imageWithText: UIImage) {
+        self.image = image
```
Comment (Brandon-T) [inline Client/Frontend/Shields/ShieldsActivityItemSourceProvider.swift:106]: Missing call to `super.init()` when inheriting `NSObject`

File: Client/Frontend/Shields/SimpleShieldsView.swift [hunk 13]
```diff
@@ -39,16 +39,39 @@ class SimpleShieldsView: UIView, Themeable {
     // Shields Up
     
     class BlockCountView: UIView, Themeable {
+        
+        private struct UX {
+            static let descriptionEdgeInset = UIEdgeInsets(top: 13, left: 16, bottom: 14, right: 16)
```
Comment (Brandon-T) [inline Client/Frontend/Shields/SimpleShieldsView.swift]: Is it 13 on top and 14 on bottom? Might as well make both 14.

File: Client/Frontend/Shields/SimpleShieldsView.swift [hunk 14]
```diff
@@ -39,16 +39,39 @@ class SimpleShieldsView: UIView, Themeable {
...
+            static let iconEdgeInset = UIEdgeInsets(top: 22, left: 14, bottom: 22, right: 14)
+            static let hitBoxEdgeInsets = UIEdgeInsets(top: -10, left: -10, bottom: -10, right: -10)
```
Comment (Brandon-T) [inline Client/Frontend/Shields/SimpleShieldsView.swift]: `UIEdgeInsets(uniform: -10.0)`

File: Client/Frontend/Shields/SimpleShieldsView.swift [hunk 15]
```diff
@@ -39,16 +39,39 @@ class SimpleShieldsView: UIView, Themeable {
...
+            static let buttonEdgeInsets = UIEdgeInsets(top: -2, left: 4, bottom: -3, right: 4)
```
Comment (Brandon-T) [inline Client/Frontend/Shields/SimpleShieldsView.swift]: -2 on top, -3 on bottom? Can we make them the same?

File: Client/Frontend/Shields/SimpleShieldsView.swift [hunk 16]
```diff
@@ -110,11 +185,23 @@ class SimpleShieldsView: UIView, Themeable {
...
-            descriptionLabel.textColor = theme.isDark ? UIColor.white : .black
-            infoButton.tintColor = theme.isDark ?
-                Colors.orange400 :
-                Colors.orange500
+            countLabel.appearanceTextColor = theme.isDark ? .white : .black
+            descriptionLabel.appearanceTextColor = theme.isDark ? .white : .black
+            
+            // Share UI only exist in locale JP
+            if Locale.current.regionCode == "JP" {
+                let contentBackgroundColor = theme.isDark ? UIColor(rgb: 0x303443) : Colors.neutral000
```
Comment (Brandon-T) [inline Client/Frontend/Shields/SimpleShieldsView.swift]: `#ColorLiteral`

**Review by Brandon-T — CHANGES_REQUESTED** [review]

File: Client/Frontend/Shields/ShieldsActivityItemSourceProvider.swift (hunk 9 above)
Comment (soner-yuksel) [reply Client/Frontend/Shields/ShieldsActivityItemSourceProvider.swift:49]: We extend the content when the application with that share type doesn't support text and image at the same time. and place the text on the bottom. So this is that extra space to place the text. You can see visually on screenshot in the description.

File: Client/Frontend/Shields/ShieldsActivityItemSourceProvider.swift (hunk 10 above)
Comment (soner-yuksel) [reply Client/Frontend/Shields/ShieldsActivityItemSourceProvider.swift]: Fix at https://github.com/brave/brave-ios/pull/3387/commits/da11867eb41296b14bf62ec79fb7344b5577df37

File: Client/Frontend/Shields/ShieldsActivityItemSourceProvider.swift (hunk 11 above)
Comment (soner-yuksel) [reply Client/Frontend/Shields/ShieldsActivityItemSourceProvider.swift:83]: Fix at https://github.com/brave/brave-ios/pull/3387/commits/da11867eb41296b14bf62ec79fb7344b5577df37

File: Client/Frontend/Shields/ShieldsActivityItemSourceProvider.swift (hunk 12 above)
Comment (soner-yuksel) [reply Client/Frontend/Shields/ShieldsActivityItemSourceProvider.swift:106]: Fix at https://github.com/brave/brave-ios/pull/3387/commits/da11867eb41296b14bf62ec79fb7344b5577df37

File: Client/Frontend/Shields/SimpleShieldsView.swift (hunk 13 above)
Comment (soner-yuksel) [reply Client/Frontend/Shields/SimpleShieldsView.swift]: Fix at https://github.com/brave/brave-ios/pull/3387/commits/da11867eb41296b14bf62ec79fb7344b5577df37

File: Client/Frontend/Shields/SimpleShieldsView.swift (hunk 14 above)
Comment (soner-yuksel) [reply Client/Frontend/Shields/SimpleShieldsView.swift]: Fix at https://github.com/brave/brave-ios/pull/3387/commits/da11867eb41296b14bf62ec79fb7344b5577df37

File: Client/Frontend/Shields/SimpleShieldsView.swift (hunk 15 above)
Comment (soner-yuksel) [reply Client/Frontend/Shields/SimpleShieldsView.swift]: Fix at https://github.com/brave/brave-ios/pull/3387/commits/da11867eb41296b14bf62ec79fb7344b5577df37

File: Client/Frontend/Shields/SimpleShieldsView.swift (hunk 16 above)
Comment (soner-yuksel) [reply Client/Frontend/Shields/SimpleShieldsView.swift]: Fix at https://github.com/brave/brave-ios/pull/3387/commits/da11867eb41296b14bf62ec79fb7344b5577df37

**Review by iccub — APPROVED** [review]
nice job :)

**Review by Brandon-T — APPROVED** [review]

**Review by jumde — APPROVED** [review]
lgtm

**Pull request merged**